core/playlistmanager.go

```diff
@@ -0,0 +1,280 @@
+		mgr.masterFullPList.Encode().Bytes())
+	if err != nil {
+		glog.Warningf("Error saving full master playlist for %s stream %v", streamID, err)
+	}
+	return mpl
+}
+
+func (mgr *BasicPlaylistManager) savePlayList(mpl *m3u8.MediaPlaylist, jobID int64, streamID StreamID,
+	name string) (string, error) {
+
+	url, err := mgr.storageDriver.SaveData(jobID, string(streamID), name,
+		mpl.Encode().Bytes())
+	if err != nil {
+		glog.Warningf("Error saving media playlist for %s stream %v", streamID, err)
+		return "", err
+	}
+	return url.Uri, err
+}
+
+// InsertLink ...
+func (mgr *BasicPlaylistManager) InsertLink(jobID int64, seqNo uint64, turi *net.TypedURI,
+	duration float64) (bool, error) {
+
+	if !drivers.IsOwnStorage(turi) {
+		return false, nil
+	}
+	if mgr.storageDriver == nil {
+		return false, nil
+	}
+	return mgr.insertLink(jobID, seqNo, turi, duration)
+}
+
+func (mgr *BasicPlaylistManager) insertLink(jobID int64, seqNo uint64, turi *net.TypedURI,
+	duration float64) (bool, error) {
+
+	tsid := StreamID(turi.StreamID)
+	if err := mgr.isRightStream(tsid); err != nil {
+		return false, err
+	}
+	mpl := mgr.getLivePL(tsid)
+	if mpl == nil {
+		mpl = mgr.createLivePL(tsid)
+	}
+	mgr.addToMediaPlaylist(turi, seqNo, duration, mpl, true)
+	mpl = mgr.getFullPL(tsid)
+	if mpl == nil {
+		mpl = mgr.createFullPL(jobID, tsid)
+	}
+	mgr.addToMediaPlaylist(turi, seqNo, duration, mpl, false)
+	_, err := mgr.savePlayList(mpl, jobID, tsid, turi.StreamID+".m3u8")
+	if err != nil {
+		glog.Warningf("Error saving full media playlist for %s stream %v", turi.StreamID, err)
+	}
+	return true, err
+}
+
+func (mgr *BasicPlaylistManager) mediaSegmentFromTuri(turi *net.TypedURI, seqNo uint64,
+	duration float64, absURL bool) *m3u8.MediaSegment {
+
+	mseg := new(m3u8.MediaSegment)
+	mseg.URI = turi.Name
+	mseg.SeqId = seqNo
+	mseg.Duration = duration
+	mseg.Title = turi.Name
+	if absURL {
+		mseg.URI = turi.Uri
+	}
+	return mseg
+}
+
+func (mgr *BasicPlaylistManager) addToMediaPlaylist(turi *net.TypedURI, seqNo uint64, duration float64,
+	mpl *m3u8.MediaPlaylist, live bool) error {
+
+	mseg := mgr.mediaSegmentFromTuri(turi, seqNo, duration, live)
+	if live && mpl.Count() >= mpl.WinSize() {
+		mpl.Remove()
+	}
+	if mpl.Count() == 0 {
+		mpl.SeqNo = mseg.SeqId
+	}
+	return mpl.AppendSegment(mseg)
+}
+
+func (mgr *BasicPlaylistManager) makeMediaSegment(seg *stream.HLSSegment, url string) *m3u8.MediaSegment {
+	mseg := new(m3u8.MediaSegment)
+	mseg.URI = url
+	mseg.Duration = seg.Duration
+	mseg.Title = seg.Name
+	mseg.SeqId = seg.SeqNo
+	return mseg
+}
+
+// GetHLSMasterPlaylist ..
+func (mgr *BasicPlaylistManager) GetHLSMasterPlaylist() *m3u8.MasterPlaylist {
+	return mgr.masterLivePList
+}
+
+// GetHLSMediaPlaylist ...
+func (mgr *BasicPlaylistManager) GetHLSMediaPlaylist(streamID StreamID) *m3u8.MediaPlaylist {
+	return mgr.mediaLiveLists[streamID]
+}
+
+// EndSession ends object storage's session
+func (mgr *BasicPlaylistManager) EndSession() {
```

Comment:
What's the difference between `EndSession` and `Cleanup` ? This `EndSession` is not used anywhere.

Reply:
Forgot to remove